contracts/AutonomousDripper.sol

```diff
@@ -0,0 +1,145 @@
+        uint64 startTimestamp,
+        uint64 durationSeconds,
+        address adminAddress,
+        uint intervalSeconds,
+        address[] memory watchlistAddresses
+    ) VestingWallet(
+        beneficiaryAddress,
+        startTimestamp,
+        durationSeconds
+    ) {
+        _admin = adminAddress;
+        lastTimestamp = startTimestamp;
+        interval = intervalSeconds;
+        assetsWatchlist = watchlistAddresses;
+    }
+
+    /**
+     * @dev Setter for the list of ERC-20 token addresses to consider for
+     * releasing. Can only be called by the admin.
+     */
+    function setAssetsWatchlist(address[] calldata newAssetsWatchlist) public virtual {
+        require(_msgSender() == _admin, "AutonomousDripper: onlyAdmin");
+        assetsWatchlist = newAssetsWatchlist;
+    }
+
+    /**
+     * @dev Loop over the assetsWatchlist and check their local balance.
+     * Returns a filtered version of the assetsWatchlist for which the local
+     * balance is greater than zero.
+     * Strongly inspired by https://github.com/smartcontractkit/chainlink/blob/0adf6c0fb256d09e4fed2f2f86a2c027ae82535d/contracts/src/v0.8/upkeeps/EthBalanceMonitor.sol#L87-L116
+     */
+    function _getAssetsHeld() internal view returns (address[] memory) {
+        address[] memory _assetsHeld = new address[](assetsWatchlist.length);
+        uint256 count = 0;
+        for (uint idx = 0; idx < assetsWatchlist.length; idx++) {
+            uint256 balance = IERC20(assetsWatchlist[idx]).balanceOf(address(this));
+            if (balance > 0) {
+                _assetsHeld[count] = assetsWatchlist[idx];
+                count++;
+            }
+        }
+        if (count != assetsWatchlist.length) {
+            assembly {
+                mstore(_assetsHeld, count)
+            }
+        }
+        return _assetsHeld;
+    }
+
+    /**
+     * @dev Runs off-chain at every block to determine if the `performUpkeep`
+     * function should be called on-chain.
+     */
+    function checkUpkeep(bytes calldata) external view override returns (
+        bool upkeepNeeded, bytes memory
+    ) {
+        address[] memory assetsHeld = new address[](assetsWatchlist.length);
+        uint256 count = 0;
+
+        if ((block.timestamp - lastTimestamp) > interval) {
+            assetsHeld = _getAssetsHeld();
+            if (assetsHeld.length > 0) {
+                return (true, abi.encode(assetsHeld));
+            }
+        }
+    }
+
+    /**
+     * @dev Contains the logic that should be executed on-chain when
+     * `checkUpkeep` returns true.
+     */
+    function performUpkeep(bytes calldata performData) external override {
+        if ((block.timestamp - lastTimestamp) > interval) {
+            lastTimestamp = block.timestamp;
+            address[] memory assetsHeld = abi.decode(performData, (address[]));
+            for (uint idx = 0; idx < assetsHeld.length; idx++) {
+                if (IERC20(assetsHeld[idx]).balanceOf(address(this)) > 0) {
```

Comment:
thinking that this will be called immediately after `checkUpkeep` splits out `True` and already there has being check if the balance of the token address was greater than zero. thinking that above this L in two ways:

- seems redundant check after the reads on `checkUpkeep`

- will add a bit of extra cost of reading that again on the chainlink keeper side that we should pay for

what are your thoughts?

Reply:
https://docs.chain.link/docs/chainlink-keepers/compatible-contracts/#revalidate-performupkeep

Reply:
gotcha!  🧠